internals/controllers/subnamespace/subnamespace_init.go

```diff
@@ -0,0 +1,219 @@
+	"github.com/dana-team/hns/internals/utils"
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// init creates a namespace for the subnamespace, creates a quota object for it if needed, creates
+// default ResourceQuota and LimitRange object in the namespace, and sets annotations on the subnamespace
+func (r *SubnamespaceReconciler) init(snsParentNS, snsObject *utils.ObjectContext) (ctrl.Result, error) {
+	ctx := snsObject.Ctx
+	logger := log.FromContext(ctx)
+	logger.Info("initializing subnamespace")
+
+	snsName := snsObject.Object.GetName()
+
+	if err := ensureSNSNamespace(snsParentNS, snsObject); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create namespace for subnamespace '%s': "+err.Error(), snsName)
+	}
+	logger.Info("successfully created namespace for subnamespace", "subnamespace", snsName)
+
+	isSNSRP := utils.GetSnsResourcePool(snsObject.Object)
+	if isSNSRP == "" {
+		if err := setSNSResourcePool(snsParentNS, snsObject); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set ResourcePool label for subnamespace '%s': "+err.Error(), snsName)
+		}
+		logger.Info("successfully set ResourcePool label for subnamespace", "subnamespace", snsName)
+	}
+
+	rqFlag, err := utils.IsRq(snsObject, danav1.SelfOffset)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to compute isRq flag for subnamespace '%s': "+err.Error(), snsName)
+	}
+
+	// if the subnamespace is a regular SNS (i.e. not a ResourcePool) OR it's an upper-rp, then create a corresponding
+	// quota object for the subnamespace. The quota object can be either a ResourceQuota or a ClusterResourceQuota
+	isSNSUpperRP, _ := utils.IsUpperResourcePool(snsObject)
+	if isSNSRP == "false" || isSNSUpperRP {
+		if res, err := ensureSNSQuotaObj(snsObject, rqFlag); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to create quota object for subnamespace '%s': "+err.Error(), snsName)
+		} else if !res.IsZero() {
+			return res, nil
+		}
+		logger.Info("successfully created quota object for subnamespace", "subnamespace", snsName)
+	}
+
+	if err := ensureDefaultSNSResourceQuota(snsObject); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create default ResourceQuota object for subnamespace '%s': "+err.Error(), snsName)
+	}
+	logger.Info("successfully created default ResourceQuota object for subnamespace", "subnamespace", snsName)
+
+	if err := ensureDefaultSNSLimitRange(snsObject); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create default Limit Range object for subnamespace '%s': "+err.Error(), snsName)
+	}
+	logger.Info("successfully created default LimitRange object for subnamespace", "subnamespace", snsName)
+
+	annots := make(map[string]string)
+	displayName := utils.GetNamespaceDisplayName(snsParentNS.Object) + "/" + snsObject.Object.GetName()
+	annots[danav1.OpenShiftDisplayName] = displayName
+	annots[danav1.DisplayName] = displayName
+	if rqFlag {
```

Comment:
Yeah I kept it this way to keep it consistent with how it currently is since I didn't want to accidentally break things. We can address this in a future PR and drop the `danav1.True` and `danav1.False` entirely.